src/braket/circuits/noise_model/noise_model.py

```diff
@@ -0,0 +1,409 @@
+        """
+        Applies the gate noise to return a new circuit that's the `noisy` version of the given
+        circuit.
+
+        Args:
+            circuit (Circuit): a circuit to apply `noise` to.
+            gate_noise_instructions (List[NoiseModelInstruction]): a list of gate noise
+                instructions to apply to the circuit.
+
+        Returns:
+            Circuit: A new circuit that's a `noisy` version of the passed in circuit. The targets
+            set will be populated with the list of targets in the new circuit.
+        """
+        new_circuit = Circuit()
+        for circuit_instruction in circuit.instructions:
+            new_circuit.add_instruction(circuit_instruction)
+            target_qubits = list(circuit_instruction.target)
+            for item in gate_noise_instructions:
+                if item.criteria.instruction_matches(circuit_instruction):
+                    if item.noise.fixed_qubit_count() == len(target_qubits):
+                        new_circuit.add_instruction(Instruction(item.noise, target_qubits))
+                    else:
+                        for qubit in target_qubits:
+                            new_circuit.add_instruction(Instruction(item.noise, qubit))
+        for result_type in circuit.result_types:
+            new_circuit.add_result_type(result_type)
+        return new_circuit
+
+    @classmethod
+    def _apply_init_noise(
+        cls,
+        circuit: Circuit,
+        init_noise_instructions: List[NoiseModelInstruction],
+    ) -> Circuit:
+        """
+        Applies the initialization noise of this noise model to a circuit and returns the circuit.
+
+        Args:
+            circuit (Circuit): A circuit to apply `noise` to.
+            init_noise_instructions (List[NoiseModelInstruction]): A list of initialization noise
+                model instructions.
+
+        Returns:
+            Circuit: The passed in circuit, with the initialization noise applied.
+        """
+        if not init_noise_instructions:
+            return circuit
+        for item in init_noise_instructions:
+            qubits = item.criteria.qubit_intersection(circuit.qubits)
+            if len(qubits) > 0:
+                circuit.apply_initialization_noise(item.noise, list(qubits))
+        return circuit
+
+    @classmethod
+    def _apply_readout_noise(
+        cls,
+        circuit: Circuit,
+        readout_noise_instructions: List[NoiseModelInstruction],
+    ) -> Circuit:
+        """
+        Applies the readout noise of this noise model to a circuit and returns the circuit.
+
+        Args:
+            circuit (Circuit): A circuit to apply `noise` to.
+            readout_noise_instructions (List[NoiseModelInstruction]): The list of readout noise
+                to apply.
+
+        Returns:
+            Circuit: The passed in circuit, with the readout noise applied.
+        """
+        if not readout_noise_instructions or not circuit.result_types:
```

Comment:
If the customer doesn't specify any result types, the default behavior is to measure all qubits. In such a case, how should the customers model this readout noise ?

Reply:
After a few iterations - we had issues dealing with edge cases with different types of result types - we decided that customers will need to add a specific result type if they wanted to add noise to it.